TestingApp/VisitorInfo/VisitorInfoModel.swift

```diff
@@ -0,0 +1,468 @@
+            self?.loadVisitorInfo()
+        }
+
+        let cancelTap = Cmd { [weak self] in
+            self?.delegate?(.cancel)
+        }
+
+        let updateTap = Cmd { [weak self] in
+            self?.updateVisitorInfo()
+        }
+
+        let hasChanges: Bool = !Self.diffBetweenVisitorInfoUpdates(
+            updateA: self.visitorInfoUpdate,
+            updateB: self.previousVisitorInfoUpdate
+        ).isEmpty
+
+        let updateButton: UpdateButton
+        switch self.updateInfoRequestState {
+        case .none, .result:
+            updateButton = .normal(updateTap, isEnabled: hasChanges)
+        case .inFlight:
+            updateButton = .loading
+        }
+
+        let props = Props(
+            title: "Visitor Info",
+            sections: [
+                fieldsSection,
+                noteSection,
+                customAttributesSection
+            ],
+            pulledToRefresh: pulledToRefresh,
+            showsRefreshing: self.fetchInfoRequestState == .inFlight,
+            cancelTap: cancelTap,
+            updateButton: updateButton
+        )
+
+        return props
+    }
+
+    func requestPropsRendered() {
+        self.delegate?(.renderProps(self.props()))
+    }
+
+    fileprivate enum DiffField {
+        case name, email, phone, note, noteUpdateMethod, customAttributesUpdateMethod
+        case customAttributes(lhs: [String: String], rhs: [String: String])
+    }
+
+    fileprivate static func diffBetweenVisitorInfoUpdates(
+        updateA: VisitorInfoUpdate,
+        updateB: VisitorInfoUpdate
+    ) -> [DiffField] {
+        var fields: [DiffField] = []
+
+        if updateA.name ?? "" != updateB.name ?? "" {
+            fields.append(.name)
+        }
+
+        if updateA.email ?? "" != updateB.email ?? "" {
+            fields.append(.email)
+        }
+
+        if updateA.phone ?? "" != updateB.phone ?? "" {
+            fields.append(.phone)
+        }
+
+        if updateA.note ?? "" != updateB.note ?? "" {
+            fields.append(.note)
+        }
+
+        if updateA.customAttributes ?? [:] != updateB.customAttributes ?? [:] {
+            fields.append(
+                .customAttributes(
+                    lhs: updateA.customAttributes ?? [:],
+                    rhs: updateB.customAttributes ?? [:]
+                )
+            )
+        }
+
+        if updateA.customAttributesUpdateMethod != updateB.customAttributesUpdateMethod {
+            fields.append(.customAttributesUpdateMethod)
+        }
+
+        if updateA.noteUpdateMethod != updateB.noteUpdateMethod {
+            fields.append(.noteUpdateMethod)
+        }
+
+        return fields
+    }
+}
+
+extension VisitorInfoModel {
+    struct Environment {
+        let fetchVisitorInfo: (@escaping (Result<GliaCore.VisitorInfo, Error>) -> Void) -> Void
+        let updateVisitorInfo: (VisitorInfoUpdate, @escaping (Result<Bool, Error>) -> Void) -> Void
+        let uuid: () -> UUID
```

Comment:
would we like to use Combine instead of functions here?

Reply:
Not sure if wrapping it into Combine publisher will be beneficial.
As side note: I tried to use Combine for reporting updated 'Props' using `combineLatest` but that did not work for me as I wanted to, so I left Combine out.